LNL/SDW AOIC: major regression with PIO commands

from Intel daily tests #/result/planresultdetail/39251?model=LNLM_SDW_AIOC&testcase=check-sof-logger

````
[ 4685.789060] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to reset HDA controller gctl 0x1
[ 4685.789067] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to reset controller during suspend
[ 4685.789070] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to power down DSP during suspend -5
[ 4685.789076] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: can't suspend (snd_sof_runtime_suspend [snd_sof] returned -5)
[ 4688.184905] kernel: soundwire_intel soundwire_intel.link.0: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[ 4688.184915] kernel: soundwire_intel soundwire_intel.link.0: intel_resume_runtime: power_up failed after bus reset: -5
[ 4688.184919] kernel: soundwire_intel soundwire_intel.link.0: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.0: -5
[ 4688.184923] kernel:  SDW0-Playback: ASoC: error at __soc_pcm_open on SDW0-Playback: -5
[ 4688.184926] kernel:  Jack Out: ASoC: error at dpcm_be_dai_startup on Jack Out: -5
[ 4688.184928] kernel:  Jack Out: ASoC: error at dpcm_fe_dai_startup on Jack Out: -5
[ 4688.186628] kernel: soundwire_intel soundwire_intel.link.0: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.0: -22
[ 4688.191686] kernel: soundwire_intel soundwire_intel.link.2: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[ 4688.191701] kernel: soundwire_intel soundwire_intel.link.2: intel_resume_runtime: power_up failed after bus reset: -5
[ 4688.191707] kernel: soundwire_intel soundwire_intel.link.2: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.2: -5
[ 4688.191713] kernel:  SDW2-Playback: ASoC: error at __soc_pcm_open on SDW2-Playback: -5
[ 4688.191717] kernel:  Speaker: ASoC: error at dpcm_be_dai_startup on Speaker: -5
[ 4688.191721] kernel:  Speaker: ASoC: error at dpcm_fe_dai_startup on Speaker: -5
[ 4688.193344] kernel: soundwire_intel soundwire_intel.link.0: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.0: -22
[ 4688.195635] kernel: soundwire_intel soundwire_intel.link.2: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.2: -22
[ 4688.200113] kernel: soundwire_intel soundwire_intel.link.1: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[ 4688.200116] kernel: soundwire_intel soundwire_intel.link.1: intel_resume_runtime: power_up failed after bus reset: -5
[ 4688.200119] kernel: soundwire_intel soundwire_intel.link.1: ASoC: error at snd_soc_pcm_component_pm_runtime_get on soundwire_intel.link.1: -5
[ 4688.200121] kernel:  SDW1-Capture: ASoC: error at __soc_pcm_open on SDW1-Capture: -5
[ 4688.200123] kernel:  Microphone: ASoC: error at dpcm_be_dai_startup on Microphone: -5
[ 4688.200124] kernel:  Microphone: ASoC: error at dpcm_fe_dai_startup on Microphone: -5
````

That's super bad...

manual tests run on a different device jf-lnlm-rvp-sdw-1 show this

````
[  363.370649] kernel: hdaudio ehdaudio0D7: no AFG or MFG node found
[  363.370671] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: device init failed for hdac device
[  367.530144] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to reset HDA controller gctl 0x1
[  367.530180] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to reset controller during suspend
[  367.530196] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: error: failed to power down DSP during suspend -5
[  367.530228] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: can't suspend (snd_sof_runtime_suspend [snd_sof] returned -5)
[  373.581625] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc4_tx_msg_unlocked: ipc message send for 0x44000000|0x31400008 failed: -19
[  373.581627] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: sof_ipc4_set_get_data: large config set failed at offset 0: -19
[  373.671146] kernel: soundwire_intel soundwire_intel.link.0: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[  373.671168] kernel: soundwire_intel soundwire_intel.link.0: intel_resume failed: -5
[  373.671178] kernel: soundwire_intel soundwire_intel.link.0: PM: dpm_run_callback(): pm_generic_resume+0x0/0x30 returns -5
[  373.671215] kernel: soundwire_intel soundwire_intel.link.0: PM: failed to resume: error -5
[  373.674568] kernel: soundwire_intel soundwire_intel.link.1: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[  373.674570] kernel: soundwire_intel soundwire_intel.link.1: intel_resume failed: -5
[  373.674571] kernel: soundwire_intel soundwire_intel.link.1: PM: dpm_run_callback(): pm_generic_resume+0x0/0x30 returns -5
[  373.674575] kernel: soundwire_intel soundwire_intel.link.1: PM: failed to resume: error -5
[  373.677853] kernel: soundwire_intel soundwire_intel.link.2: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[  373.677855] kernel: soundwire_intel soundwire_intel.link.2: intel_resume failed: -5
[  373.677856] kernel: soundwire_intel soundwire_intel.link.2: PM: dpm_run_callback(): pm_generic_resume+0x0/0x30 returns -5
[  373.677858] kernel: soundwire_intel soundwire_intel.link.2: PM: failed to resume: error -5
[  373.681192] kernel: soundwire_intel soundwire_intel.link.3: intel_link_power_up: hdac_bus_eml_sdw_power_up failed: -5
[  373.681193] kernel: soundwire_intel soundwire_intel.link.3: intel_resume failed: -5
[  373.681194] kernel: soundwire_intel soundwire_intel.link.3: PM: dpm_run_callback(): pm_generic_resume+0x0/0x30 returns -5
[  373.681196] kernel: soundwire_intel soundwire_intel.link.3: PM: failed to resume: error -5
[  373.787757] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: IP_MCP_CONTROL_SW_RST is not cleared
[  373.789266] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: MCP_CONTROL_CMD_RST is not cleared
[  373.789363] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: MCP_CONTROL_SOFT_RST is not cleared
[  373.790132] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: MCP_CONTROL_CLK_STOP_CLR is not cleared
[  373.790136] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: MCP_CONFIG_UPDATE_BIT is not cleared
[  373.790138] kernel: soundwire_intel soundwire_intel.link.3: sdw_cdns_clock_stop failed: MCP_CONTROL_HW_RST is not cleared
[  373.790155] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: IP_MCP_CONTROL_SW_RST is not cleared
[  373.790159] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: MCP_CONTROL_CMD_RST is not cleared
[  373.790160] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: MCP_CONTROL_SOFT_RST is not cleared
[  373.790162] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: MCP_CONTROL_CLK_STOP_CLR is not cleared
[  373.790175] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: MCP_CONFIG_UPDATE_BIT is not cleared
[  373.790177] kernel: soundwire_intel soundwire_intel.link.2: sdw_cdns_clock_stop failed: MCP_CONTROL_HW_RST is not cleared
[  373.790189] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: IP_MCP_CONTROL_SW_RST is not cleared
[  373.790192] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: MCP_CONTROL_CMD_RST is not cleared
[  373.790194] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: MCP_CONTROL_SOFT_RST is not cleared
[  373.790195] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: MCP_CONTROL_CLK_STOP_CLR is not cleared
[  373.790198] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: MCP_CONFIG_UPDATE_BIT is not cleared
[  373.790209] kernel: soundwire_intel soundwire_intel.link.1: sdw_cdns_clock_stop failed: MCP_CONTROL_HW_RST is not cleared
[  373.790220] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: IP_MCP_CONTROL_SW_RST is not cleared
[  373.790245] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: MCP_CONTROL_CMD_RST is not cleared
[  373.790247] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: MCP_CONTROL_SOFT_RST is not cleared
[  373.790249] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: MCP_CONTROL_CLK_STOP_CLR is not cleared
[  373.790251] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: MCP_CONFIG_UPDATE_BIT is not cleared
[  373.790253] kernel: soundwire_intel soundwire_intel.link.0: sdw_cdns_clock_stop failed: MCP_CONTROL_HW_RST is not cleared
````

which is even worse - something is clearly not quite right.

We'll need to bisect this one further, I don't recall seeing all this before. @fredoh9 @marc-hb @bardliao @kv2019i @ujfalusi FYI

I went back to the last PR I tested on this platform, which was precisely aimed at fixing kmod-load-unload issues
https://github.com/thesofproject/linux/pull/4849 commit 32f632b22d212ab3b70b1cdadd507735f7569f6c

With this branch, I don't see any error proper at the kernel level. The sof-test do fail however, but I have no explanation for the results.
````
declare -- cmd="journalctl_cmd --since=@1711729127"
2024-03-29 16:18:49 UTC [INFO] checking if firmware is loaded successfully
Polling 'sof_firmware_boot_complete --since=@1711729127' every 1s for 3s
2024-03-29 16:18:52 UTC [ERROR] Failed to load firmware after module insertion
2024-03-29 16:18:52 UTC [INFO] Starting func_exit_handler(1)
2024-03-29 16:18:52 UTC [ERROR] Starting func_exit_handler(), exit status=1, FUNCNAME stack:
2024-03-29 16:18:52 UTC [ERROR]  die()  @  /root/sof-test/case-lib/lib.sh
2024-03-29 16:18:52 UTC [ERROR]  main()  @  /root/sof-test/test-case/check-kmod-load-unload.sh:97
2024-03-29 16:18:52 UTC [INFO] pkill -TERM -f mtrace-reader.py
2024-03-29 16:18:52 UTC [ERROR] mtrace-reader.py was already dead
2024-03-29 16:18:52 UTC [INFO] ktime=511 sof-test PID=2665: ending
2024-03-29 16:18:52 UTC [INFO] Test Result: FAIL!
````

[log.txt](https://github.com/thesofproject/linux/files/14806294/log.txt)

--- a/sound/soc/sof/intel/hda-ctrl.c
+++ b/sound/soc/sof/intel/hda-ctrl.c
@@ -188,6 +188,7 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev)
 	struct hdac_bus *bus = sof_to_bus(sdev);
 	struct hdac_stream *stream;
 	int sd_offset, ret = 0;
+	u32 gctl;
 
 	if (bus->chip_init)
 		return 0;
@@ -195,6 +196,12 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev)
 	hda_codec_set_codec_wakeup(sdev, true);
 
 	hda_dsp_ctrl_misc_clock_gating(sdev, false);
+
+	/* clear WAKE_STS if not in reset */
+	gctl = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, SOF_HDA_GCTL);
+	if (gctl & SOF_HDA_GCTL_RESET)
+		snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR,
+				  SOF_HDA_WAKESTS, SOF_HDA_WAKESTS_INT_MASK);
 
 	/* reset HDA controller */
 	ret = hda_dsp_ctrl_link_reset(sdev, true);

